internal/sys/vault.go

```diff
@@ -6,48 +6,153 @@ package sys
+
+// Seal seals the Vault. Once sealed, any subsequent Vault operation,
+// returns ErrSealed until the Vault gets unsealed again.
+func (v *Vault) Seal(ctx context.Context) error {
+	if v.sealed {
+		return kes.ErrSealed
+	}
+	if err := v.fs.Seal(ctx); err != nil {
+		return err
+	}
+	v.admin = ""
+	v.enclaves = map[string]*Enclave{}
+	v.sealed = true
+	return nil
+}
+
+// Unseal unseals the Vault. In case of an unsealed Vault,
+// Unseal is a no-op.
+func (v *Vault) Unseal(ctx context.Context, keys ...UnsealKey) error {
+	if !v.sealed {
+		return nil
+	}
+	if err := v.fs.Unseal(ctx, keys...); err != nil {
+		return err
+	}
+	admin, err := v.fs.Admin(ctx)
+	if err != nil {
+		return err
+	}
+	v.admin = admin
+	v.enclaves = map[string]*Enclave{}
+	v.sealed = false
+	return nil
+}
+
+// Admin returns the current Vault admin identity.
+func (v *Vault) Admin(ctx context.Context) (kes.Identity, error) {
+	if !v.admin.IsUnknown() {
+		return v.admin, nil
+	}
+
+	admin, err := v.fs.Admin(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	v.cacheLock.Lock()
+	v.admin = admin
+	v.cacheLock.Unlock()
+	return v.admin, nil
+}
 
-// A Vault manages a set of Enclaves.
+// CreateEnclave creates a new enclave with the given name and
+// enclave admin identity.
 //
-// It is either in a sealed or unsealed state. When the
-// Vault is sealed it does not process any requests except
-// unseal requests. Once unsealed, Vault provides access
-// to existing enclaves.
-type Vault interface {
-	// Seal seals the Vault. Once sealed, any subsequent operation
-	// returns ErrSealed.
-	//
-	// It returns ErrSealed if the Vault is already sealed.
-	Seal(ctx context.Context) error
-
-	// Unseal unseals the Vault.
-	//
-	// It returns no error If the Vault is already unsealed.
-	Unseal(ctx context.Context, keys ...UnsealKey) error
-
-	// Operator returns the identity of the Vault operator.
-	SysAdmin(context.Context) (kes.Identity, error)
-
-	// CreateEnclave creates and returns a new Enclave if and only if
-	// no Enclave with the given name exists.
-	//
-	// It returns ErrEnclaveExists if an Enclave with the given name
-	// already exists.
-	CreateEnclave(ctx context.Context, name string, admin kes.Identity) (*EnclaveInfo, error)
-
-	// GetEnclave returns the Enclave associated with the given name.
-	//
-	// It returns ErrEnclaveNotFound if no Enclave with the given
-	// name exists.
-	GetEnclave(ctx context.Context, name string) (*Enclave, error)
-
-	// DeleteEnclave deletes the Enclave with the given name.
-	DeleteEnclave(ctx context.Context, name string) error
+// It returns ErrEnclaveExists if such an enclave already exists.
+func (v *Vault) CreateEnclave(ctx context.Context, name string, admin kes.Identity) (*EnclaveInfo, error) {
+	if name == "" {
+		name = DefaultEnclaveName
+	}
+
+	if v.sealed {
+		return nil, kes.ErrSealed
+	}
+	if admin.IsUnknown() {
+		// TODO: disallow
+	}
+	if admin == v.admin {
+		// TODO: disallow
+	}
```

Comment:
are these ones gonna be implemented in the future?

Reply:
There are already. Check the latest revision.